FOCUS #165: Pricing Quantity (#225)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/metrics/pricingquantity.md

# Pricing Quantity

> The Pricing Quantity represents the volume of a given resource and/or service used or purchased based on the [Pricing Unit](#pricingunit). Distinct from [Usage Quantity](#usagequantity) (complementary to [Usage Unit](#usageunit)), it focuses on pricing and cost, not resource and service consumption.

Reviewer: Update to Pricing Measure (and link)

Reviewer 2: resolved

> The Pricing Quantity represents the volume of a given resource and/or service used or purchased based on the [Pricing Measure](#pricinmeasure). Distinct from [Usage Quantity](#usagequantity) (complementary to [Usage Unit](#usageunit)), it focuses on pricing and cost, not resource and service consumption.

Reviewer: ```suggestion
The Pricing Quantity represents the volume of a given resource and/or service used or purchased based on the [Pricing Measure](#pricingmeasure). Distinct from [Usage Quantity](#usagequantity) (complementary to [Usage Unit](#usageunit)), it focuses on pricing and cost, not resource and service consumption.
```

> The Pricing Quantity represents the volume of a given resource and/or service used or purchased based on the [Pricing Measure](#pricingmeasure). Distinct from [Usage Quantity](#usagequantity) (complementary to [Usage Unit](#usageunit)), it focuses on pricing and cost, not resource and service consumption.

Reviewer 3: Does it make sense to include something that explains how Pricing Quantity * Pricing Unit = <insert cost column here>?

> The Pricing Quantity represents the volume of a given resource and/or service used or purchased based on the [Pricing Measure](#pricingmeasure). Distinct from [Usage Quantity](#usagequantity) (complementary to [Usage Unit](#usageunit)), it focuses on pricing and cost, not resource and service consumption.

Reviewer 4: > a given resource and/or service used or purchased

I wonder whether we can specifically refer to the SKU concept: "a given SKU used or purchased".

Reviewer 2: I believe we purchase resources and/or services, and the underlying SKUs (one or many) are metered and charged.

> PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [ChargeType](#chargetype) is "Adjustment". This column MUST NOT contain null values where unit prices are not null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.

Reviewer 2: ```suggestion
PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [ChargeCategory](#chargecategory) is "Adjustment". This column MUST NOT contain null values where unit prices are not null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.
```

Reviewer 2: Do we need to be explicit about ChargeSubcategory: Credit? Should we add something like this: 
`...and MUST be negative in cases where [ChargeSubcategory](#chargesubcategory) is 'Credit'`

> PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [ChargeType](#chargetype) is "Adjustment". This column MUST NOT contain null values where unit prices are not null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.

Reviewer 4: Minor suggestion (feel free to resolve this comment) about:

>  multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.

The FOCUS names of unit price and cost metrics could be also explicitly mentioned to ease user understanding.

Reviewer 2: The math is applicable to all unit price and cost metrics. In version 1.0, we only have ListUnitPrice, and in the future, we'll likely introduce NegotiatedUnitPrice (name subject to change). The idea is to avoid changing this spec when/if we add additional columns or rename existing ones.

> PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [ChargeCategory](#chargecategory) is "Adjustment". This column MUST NOT contain null values where unit prices are not null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.

Reviewer 2: ```suggestion
PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [ChargeCategory](#chargecategory) is "Adjustment" and MUST be negative in cases where [ChargeSubcategory](#chargesubcategory) is "Credit". This column MUST NOT contain null values where unit prices are not null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.
```

Reviewer 2: Not sure if we should impose negative values in case of credit.

Reviewer: I would think in the case of ad-hoc credits, we would not want to do this. Putting a negative quantity implies that we have a rate associated with the charge? which we don't since there's no SKU? Interested to understand if I'm thinking of this wrong.

Reviewer 5: Agreed, it's strange. As currently defined, enforcing a negative `pricingQuantity` for a credit would force a non-null `pricingMeasure`. Any quantity `pricingMeasure` strengthens the implication that there are rates/measures associated with all credits. Vote to not include.

Reviewer 2: Thanks for your feedback! Closing the conversation

> PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [ChargeCategory](#chargecategory) is "Adjustment". This column MUST NOT contain null values where unit prices are not null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.

Reviewer 6: ChargeCategory is currently named [ChargeType](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/dimensions/chargetype.md) and I'm not finding an open PR that changes that.
Also, there should be a space between those two words as the name of the column instead of the column ID.

```suggestion
PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [Charge Type](#chargetype) is "Adjustment". This column MUST NOT contain null values where unit prices are not null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.
```

Reviewer 6: Other columns are referred to by their name so I made the above change for consistency. PricingQuantity in this paragraph is self-referential so I'm not sure whether referring to it by the name or the ID is preferred.

Reviewer 7: We'll do a spec-wide sweep for ChargeType to Category once the PRs that reference it are in. Either is good, but it makes sense to use ChargeType in case anything comes up.

Reviewer 6: _Updated my suggestion reflect ChargeType instead of Charge Type_

Reviewer 2: @udam-f2 @flanakin should we switch everything back to ChargeType, AdjustmentType? Or leave it AS-IS?

> PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [AdjustmentType](#adjustmenttype) is 'Refund'. This column MUST NOT be null if unit prices are not null and MUST be null if unit prices are null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.

Reviewer 2: ```suggestion
PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [AdjustmentType](#adjustmenttype) is 'Refund'. This column MUST NOT be null if [SkuPriceId](#skupriceid) is not null and MUST be null if SkuPriceId is null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.
```

Reviewer 2: I prefer the original version as it aligns with the subsequent sentence referring to unit prices not being null...

Reviewer: unit prices seems not very clear to me since there may be multiple in the future (negotiated) and also, is not referring to something that can be linked to.

> PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [AdjustmentType](#adjustmenttype) is 'Refund'. This column MUST NOT be null where unit prices are not null and MUST be null if unit prices are null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.

Reviewer 2: ```suggestion
PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [AdjustmentType](#adjustmenttype) is 'Refund'. This column MUST NOT be null when [SkuPriceId](#skupriceid) is not null and MUST be null if SkuPriceId is null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.
```

> PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [AdjustmentType](#adjustmenttype) is 'Refund'. This column MUST NOT be null where unit prices are not null and MUST be null if unit prices are null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.

Reviewer 2: ```suggestion
PricingQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [AdjustmentType](#adjustmenttype) is 'Refund'. This column MUST NOT be null if unit prices are not null and MUST be null if unit prices are null. When unit prices are not null, multiplying PricingQuantity by a unit price MUST produce a result equal to the corresponding cost metric.
```
